sources/engine/Stride.Rendering/Rendering/Mesh.cs

```diff
@@ -76,5 +82,165 @@ public Mesh(MeshDraw meshDraw, ParameterCollection parameters)
+            if (containsShape)
+            {
+                Shapes[shape] = weight;
+
+                if (MATBSHAPE != null)
+                {
+                    int index=Shapes.IndexOf(c => c.Key == shape);
+                    for (int i = 0; i < Draw.VertexMapping.Length; i++)
+                    {
+                        var shiftedIndex = index * Draw.VertexMapping.Length + i;
+                        MATBSHAPE[shiftedIndex][3] = weight;
+                    }
+                }
+            }   
+        }
+
+        public bool BlendShapeProcessingNecessary = false;
+
+       public Matrix[] MATBSHAPE { get; set; }
+
+       public float BasisKeyWeight { get; set; }
+
+       public void ProcessBlendShapes()
+       {
+            if (BlendShapeProcessingNecessary)
+            {
+                if (Shapes == null || Shapes.Count < 1) { return; }
+                var blendShapeVertices = AdjustPositionToDrawInstance();
+                var blendShapesWeights = GetBlendWeights(out float cummulativeWeight); ;
+                BasisKeyWeight = 1 - cummulativeWeight;
+
+                MATBSHAPE = new Matrix[Draw.VertexMapping.Length*blendShapesWeights.Length];
+                for (int iBendShape = 0; iBendShape < blendShapesWeights.Length; iBendShape++)
+                {
+                    for (int iVert = 0; iVert < Draw.VertexMapping.Length; iVert++)
+                    {
+                        var vectorValue = blendShapeVertices.ElementAt(iBendShape * Draw.VertexMapping.Length + iVert);
+                        MATBSHAPE[iBendShape * Draw.VertexMapping.Length + iVert].Row1 = new Vector4(vectorValue, blendShapesWeights[iBendShape]);
+
+                    }
+                }
+                BlendShapeProcessingNecessary = false;
+            }
+        }
+        
+
+        public void AddBlendShapes(Shape shape, float weight)
+        {
+            (Shapes ??= new()).Add(shape, weight); 
+            Parameters.Set(MaterialKeys.HasBlendShape, true);
+            
+        }
+
+        public float[] GetBlendWeights(out float cummulativeWeight)
+        {
+            float[] shapeWeights = new float[Shapes.Count];
+            cummulativeWeight = 0f;
+             
+            for (int i = 0; i < Shapes.Count; i++)
+            {
+                var shapeKV = Shapes.ElementAt(i);
+                var shape = shapeKV.Key;
+                var shapeWeight = Math.Clamp(shapeKV.Value, 0f, 1f);
+                float adjustedWeight = shapeWeight;
+               // if (cummulativeWeight >= 1f) { adjustedWeight = 0f; }
+               // else if (cummulativeWeight + shapeWeight > 1) { adjustedWeight = 1 - cummulativeWeight; }
+                //else
+                //{
+                  //  adjustedWeight = shapeWeight;
+                //}
+                cummulativeWeight += adjustedWeight;
+                shapeWeights[i] = adjustedWeight;
+            }
```

Comment:
Best to iterate over the dictionary itself since `ElementAt` is a `On` operation. Something like 
```cs
int i = -1;
foreach(var shapeKV in Shapes)
{
   i++;
   var shapeWeight = Math.Clamp(shapeKV.Value, 0f, 1f);
   // ...
   shapeWeights[i] = adjustedWeight;
}
```